src/jit/morph.cpp

```diff
@@ -490,6 +490,11 @@ GenTreePtr Compiler::fgMorphCast(GenTreePtr tree)
                 shiftAmount      = gtFoldExpr(shiftAmount);
                 oper->gtOp.gtOp2 = shiftAmount;
 
+#if DEBUG
```

Comment:
I do not think that this workaround is a right fix.

Lets imagine that `shiftAmount` is such expression: "(a>0)? b:b", where `b` is not leaf node. `gtFoldExpr` will return `b`, it is flag will be cleared after your change, but not its children flags. So we will hit the assert again.

I think we should clean `gtFoldExpr` not to call `fgMorphDone` (It doesn't require many changes) or revert the optimization, if it doesn't worth it.

Reply:
That would be true if `gtFoldExpr` did a full top-down bottom-up traversal, but it doesn't do that. So from what I can see in cases where `gtFoldExpr` does walk down one level, `b` is only flagged if it is the ultimate result of folding and if it is a constant.

But I agree with you that not calling `fgMorphTreeDone` at all during `gtFoldxxx` is a better option. Let me look into that and see if it really is a straightforward change.

Reply:
> That would be true if gtFoldExpr did a full top-down bottom-up traversal, but it doesn't do that. So from what I can see in cases where gtFoldExpr does walk down one level, b is only flagged if it is the ultimate result of folding and if it is a constant.

You are right, than the change should work.

Reply:
If you agree, how about we take this fix and the separately work on getting `gtFoldExpr` out of the business of setting the morphed flag?

Reply:
It sound good for me.